File: go/nebula/pkg/net/Connection.go
/* Copyright (c) 2020 vesoft inc. All rights reserved.
 *
 * This source code is licensed under Apache 2.0 License,
 * attached with Common Clause Condition 1.0, found in the LICENSES directory.
 */

package nebulaNet

import (
	"fmt"

	"github.com/facebook/fbthrift/thrift/lib/go/thrift"
	conf "github.com/vesoft-inc/nebula-clients/go/nebula/pkg/conf"
	data "github.com/vesoft-inc/nebula-clients/go/nebula/pkg/data"
	"github.com/vesoft-inc/nebula-clients/go/nebula/pkg/nebula/graph"
)

type Connection struct {
	SeverAddress data.HostAddress
	graph        *graph.GraphServiceClient
}

func NewConnection(severAddress data.HostAddress) *Connection {
	return &Connection{
		SeverAddress: severAddress,
		graph:        nil,
	}
}

func (cn *Connection) Open(hostAddress data.HostAddress, conf conf.PoolConfig) (err error) {
	ip := hostAddress.Host
	port := hostAddress.Port
	newAdd := fmt.Sprintf("%s:%d", ip, port)
	timeoutOption := thrift.SocketTimeout(conf.TimeOut)
	addressOption := thrift.SocketAddr(newAdd)
	sock, err := thrift.NewSocket(timeoutOption, addressOption)
	if err != nil {
		return fmt.Errorf("Failed to create a net.Conn-backed Transport,: %s", err.Error())
	}

	transport := thrift.NewBufferedTransport(sock, 128<<10)
	pf := thrift.NewBinaryProtocolFactoryDefault()
	cn.graph = graph.NewGraphServiceClientFactory(transport, pf)

	if err = cn.graph.Transport.Open(); err != nil {
		return fmt.Errorf("Failed to open transport, error: %s", err.Error())
	}
	if cn.graph.Transport.IsOpen() == false {
		return fmt.Errorf("Transport is off")
	}
	return nil
}

// Authenticate
func (cn *Connection) Authenticate(username, password string) (*graph.AuthResponse, error) {
	resp, err := cn.graph.Authenticate([]byte(username), []byte(password))
	if err != nil {
		err = fmt.Errorf("Authentication fails, %s", err.Error())
		if e := cn.graph.Close(); e != nil {
			err = fmt.Errorf("Fail to close transport, error: %s", e.Error())
		}
		return nil, err
	}
	return resp, err
}

func (cn *Connection) Execute(sessionID int64, stmt string) (*graph.ExecutionResponse, error) {
	return cn.graph.Execute(sessionID, []byte(stmt))
}

// unsupported
// func (client *GraphClient) ExecuteJson((sessionID int64, stmt string) (*graph.ExecutionResponse, error) {
// 	return cn.graph.ExecuteJson(sessionID, []byte(stmt))
// }

// Check connection to host address
func (cn *Connection) Ping() bool {
	_, err := cn.Execute(1, "YIELD 1")
	if err != nil {
		return false
	}
	return true
}

// Sign out and release seesin ID
func (cn *Connection) SignOut(sessionID int64) error {
	// Release session ID to graphd
	if err := cn.graph.Signout(sessionID); err != nil {
		return err
	}
	return nil
}

// Close transport
func (cn *Connection) Close() {
	cn.graph.Close()
}

func IsError(resp *graph.ExecutionResponse) bool {
	return resp.GetErrorCode() != graph.ErrorCode_SUCCEEDED
}

File: cpp/src/client/Connection.cpp
/* Copyright (c) 2020 vesoft inc. All rights reserved.
 *
 * This source code is licensed under Apache 2.0 License,
 * attached with Common Clause Condition 1.0, found in the LICENSES directory.
 */

#include <thrift/lib/cpp/async/TAsyncSocket.h>
#include <thrift/lib/cpp2/async/HeaderClientChannel.h>

#include "common/interface/gen-cpp2/GraphServiceAsyncClient.h"
#include "nebula/client/Connection.h"

namespace nebula {

Connection::Connection() : client_{nullptr} {}

Connection::~Connection() {
    delete client_;
}

Connection &Connection::operator=(Connection &&c) {
    delete client_;
    client_ = c.client_;
    c.client_ = nullptr;
    return *this;
}

bool Connection::open(const std::string &address, int32_t port) {
    try {
        auto socket = apache::thrift::async::TAsyncSocket::newSocket(
            folly::EventBaseManager::get()->getEventBase(), address, port, 0 /*TODO(shylock) pass from config*/);

        client_ = new graph::cpp2::GraphServiceAsyncClient(
            apache::thrift::HeaderClientChannel::newChannel(socket));
    } catch (const std::exception &) {
        return false;
    }
    return true;
}

AuthResponse Connection::authenticate(const std::string &user, const std::string &password) {
    if (client_ == nullptr) {
        return AuthResponse{
            ErrorCode::E_DISCONNECTED, nullptr, std::make_unique<std::string>("Not open connection.")};
    }

    AuthResponse resp;
    try {
        client_->sync_authenticate(resp, user, password);
    } catch (const std::exception &ex) {
        return AuthResponse{
            ErrorCode::E_RPC_FAILURE, nullptr, std::make_unique<std::string>("Unavailable Connection.")};
    }
    return resp;
}

ExecutionResponse Connection::execute(int64_t sessionId, const std::string &stmt) {
    if (client_ == nullptr) {
        return ExecutionResponse{ErrorCode::E_DISCONNECTED, 0};
    }

    ExecutionResponse resp;
    try {
        client_->sync_execute(resp, sessionId, stmt);
    } catch (const std::exception &ex) {
        return ExecutionResponse{ErrorCode::E_RPC_FAILURE, 0};
    }

    return resp;
}

void Connection::asyncExecute(int64_t sessionId, const std::string &stmt, ExecuteCallback cb) {
    if (client_ == nullptr) {
        cb(ExecutionResponse{ErrorCode::E_DISCONNECTED});
        return;
    }
    client_->future_execute(sessionId, stmt).thenValue([cb = std::move(cb)](auto &&resp) {
        cb(std::move(resp));
    });
}

std::string Connection::executeJson(int64_t sessionId, const std::string &stmt) {
    if (client_ == nullptr) {
        // TODO handle error
        return "";
    }

    std::string json;
    try {
        client_->sync_executeJson(json, sessionId, stmt);
    } catch (const std::exception &ex) {
        // TODO handle error
        return "";
    }

    return json;
}

void Connection::asyncExecuteJson(int64_t sessionId,
                                  const std::string &stmt,
                                  ExecuteJsonCallback cb) {
    if (client_ == nullptr) {
        cb("");
        return;
    }
    client_->future_executeJson(sessionId, stmt).thenValue(std::move(cb));
}

bool Connection::isOpen() {
    return ping();
}

void Connection::close() {
    if (client_ != nullptr) {
        static_cast<apache::thrift::ClientChannel *>(client_->getChannel())->closeNow();
    }
}

bool Connection::ping() {
    auto resp = execute(-1 /*Only check connection*/, "YIELD 1");
    if (resp.error_code == ErrorCode::E_RPC_FAILURE ||
        resp.error_code == ErrorCode::E_DISCONNECTED) {
        return false;
    }
    return true;
}

void Connection::signout(int64_t sessionId) {
    if (client_ != nullptr) {
        client_->sync_signout(sessionId);
    }
}

}   // namespace nebula
